include/xtensor/xsort.hpp

```diff
@@ -362,6 +355,402 @@ namespace xt
         }
     }
 
+    /************************************************
+     * Implementation of partition and argpartition *
+     ************************************************/
+
+    /**
+     * Partially sort xexpression
+     *
+     * Partition shuffles the xexpression in a way so that the kth element
+     * in the returned xexpression is in the place it would appear in a sorted
+     * array and all elements smaller than this entry are placed (unsorted) before.
+     *
+     * The optional third parameter can either be an axis or ``xnone()`` in which case
+     * the xexpression will be flattened.
+     *
+     * This function uses ``std::nth_element`` internally.
+     *
+     * \code{cpp}
+     * xt::xarray<float> a = {1, 10, -10, 123};
+     * std::cout << xt::partition(a, 0) << std::endl; // {-10, 1, 123, 10} the correct entry at index 0
+     * std::cout << xt::partition(a, 3) << std::endl; // {1, 10, -10, 123} the correct entry at index 3
+     * std::cout << xt::partition(a, {0, 3}) << std::endl; // {-10, 1, 10, 123} the correct entries at index 0 and 3
+     * \endcode
+     *
+     * @param e input xexpression
+     * @param kth_container a container of ``indices`` that should contain the correctly sorted value
+     * @param axis either integer (default = -1) to sort along last axis or ``xnone()`` to flatten before sorting
+     *
+     * @return partially sorted xexpression
+     */
+    template <class E, class C, class R = typename detail::flatten_sort_result_type<E>::type,
```

Comment:
I think that's not really applicable here because this is just sorting, not reducing. And this only selects a flat container of the same size + 1 dimension.

Reply:
Ok, sorry, just found out the reason for this ... it's that we have an optimization when calling partition with a 1D container:

https://github.com/QuantStack/xtensor/pull/991/files/d94b41852a5925f1ea3bf79c6e6c5863bd5622b5#diff-f19e17acb408924db2b15e7d68aa6b93R442

In that case we just force the return type to something in order to avoid compilation problems with mismatching return types.

Should we keep it like this?